### PR TITLE
Server upgrades

### PR DESCRIPTION
**Pull request by kashishgrover**
This is a task branch for https://github.com/Hashnode/mern-starter/issues/344

### CODE REVIEW

File: server/controllers/post.controller.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Get all posts
+ * @param req
+ * @param res
+ * @returns void
+ */
+exports.getPosts = function (req, res) {
+  Post.find().sort('-dateAdded').exec((err, posts) => {
+    if (err) {
+      res.status(500).send(err);
```
Comment (spsiddarthan) [inline server/controllers/post.controller.js]: Let's add early returns whenever there's an error.

File: server/controllers/post.controller.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Get a single post
+ * @param req
+ * @param res
+ * @returns void
+ */
+exports.getPost = function (req, res) {
+  Post.findOne({ cuid: req.params.cuid }).exec((err, post) => {
+    if (err) {
```
Comment (spsiddarthan) [inline server/controllers/post.controller.js]: `if (err !! !post)`

File: server/controllers/post.controller.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * @param res
+ * @returns void
+ */
+exports.deletePost = function (req, res) {
+  Post.findOne({ cuid: req.params.cuid }).exec((err, post) => {
+    if (err) {
+      res.status(500).send(err);
+    }
+
+    post.remove(() => {
```
Comment (spsiddarthan) [inline server/controllers/post.controller.js]: Error handling in the callback is necessary.

File: server/index.js [hunk 4]
```diff
@@ -1,10 +1,40 @@
...
 const path = require("path");
+const bodyParser = require("body-parser");
+const mongoose = require("mongoose");
 
+//Initialize Express App
 const app = express();
 
+//Local Imports
+const serverConfig = require("./config");
+const posts = require('./routes/post.routes');
```
Comment (abinavseelan) [inline server/index.js]: Maybe we can have a single require for routes in general? 🙂 

Have an `index.js` in the the `/routes` folder which aggregates all the routes and in `server/index.js` we can just `require('/routes);` and use it across the server file?

For example
```js
const routes = require('./routes');

...

app.use('/api/' , routes.post);
```

File: server/controllers/post.controller.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
+const Post = require('../models/post');
+const cuid = require('cuid');
+const slug = require('limax');
```
Comment (abinavseelan) [inline server/controllers/post.controller.js]: Do we need to support `cuid` and `limax`? 🤔

These packages make sense only if the user wants to build a blog, taking after our demo project right? The remaining section of users will have two libraries being installed that would not be used.

Comment (abinavseelan) [reply server/controllers/post.controller.js]: I feel the out-of-the-box mongo `_id` can suffice. Also, forgot to mention the `sanitize-html` package as well. 🙂

Comment (kashishgrover) [reply server/controllers/post.controller.js]: makes sense. Let's keep cuid, however. We use it in delete and get post by cuid api's 🙂

**Comment by kashishgrover** [comment]
Cheers 😄 

screen shot 2018-01-29 at 1 35 07 pm

File: server/index.js [hunk 6]
```diff
@@ -30,19 +30,21 @@ mongoose.connect(isTest ? serverConfig.testMongoURL : serverConfig.mongoURL, (er
...
 app.use('/api', routes.posts);
 
-var server = app.listen(serverConfig.port, (error) => {
-  if (!error) {
-    console.log(`MERN is running on port: ${serverConfig.port}! Build something amazing!`); // eslint-disable-line
-  }
-});
+if (!isTest) {
+  // Testing does not require you to listen on a port
+  app.listen(serverConfig.port, (error) => {
```
Comment (kashishgrover) [inline server/index.js]: oh wow 👍

**Comment by abinavseelan** [comment]
Great job @kashishgrover 🎉

**Pull request merged**